File: 0005-最长回文子串/Solution.java
// 正序和反序看都一样即为回文串
class Solution {
    public String longestPalindrome(String s) {
        if(s == null || s.length() == 0){
            return "";
        }
        // dp[i][j]->代表i~j子串是否为回文串
        // dp[i][j]=dp[i+1][j-1] && Si==Sj
        boolean[][] dp = new boolean[s.length()][s.length()];
        int maxlength = 1;
        int index = 0;
        int templength = 0;
        for(int i=0; i<s.length(); i++){
            dp[i][i] = true;
        }
        for(int i=s.length()-2; i>=0; i--){
            for(int j=i+1; j<s.length(); j++){
                if(s.charAt(i) != s.charAt(j)){
                    dp[i][j] = false;
                    continue;
                }
                if(i+1 == j || dp[i+1][j-1]){
                    dp[i][j] = true;
                    templength = j-i+1;
                    if(templength > maxlength){
                        maxlength = templength;
                        index = i;
                    }
                }
            }
        }                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        
        return s.substring(index, maxlength+index);
    }
}

File: 0005-最长回文子串/solution.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        dp = [[False] * len(s) for _ in range(len(s))]
        for i in range(len(dp)):
            dp[i][i] = True
        begin, end = 0, 0
        for i in range(len(s)-1, -1, -1):
            for j in range(i+1, len(s)):
                if j == i+1:
                    dp[i][j] = s[i] == s[j]
                else:
                    dp[i][j] = (s[i] == s[j]) and (dp[i+1][j-1])
                if dp[i][j] and j-i >= end-begin:
                    begin, end = i, j
        return s[begin:end+1]

    def longestPalindrome(self, s: str) -> str:
        ret = ""
        if not s:
            return ret
        for i in range(len(s)):
            for j in range(i, len(s)):
                tmp = s[i:j+1]
                if tmp == tmp[::-1] and len(tmp) > len(ret):
                    ret = tmp
        return ret
